### ISSUE TITLE
Allow interpolation of secrets

### ISSUE THREAD

**Issue by rdpa** [jenkinsci/configuration-as-code-plugin#697]
- Jenkins version: LTS Docker (2.150.1)
- Plugin version: 1.4
- OS: Alpine 3.8 in Docker

Currently, secrets are only injected if they are the entire value of the key value pair. For example, the following would not get interpolated:
```
...
    someKey: The value of SECRET_VAR is ${SECRET_VAR}
...
```

**Comment by imalik-rms** [comment]
i think there is an open PR for that https://github.com/jenkinsci/configuration-as-code-plugin/pull/690

**Comment by rdpa** [comment]
I don't think so. Looking at the code changes, that PR is specifically in the context of SeedJobConfigurator. This feature would work in any part of the config.

**Comment by dgarzon** [comment]
@rdpa This is something I started to take a look at today. The issue lies in: `SecretSource` and the use of a `Pattern`. If I have time, I will try to implement something this week.

**Comment by antgel** [comment]
I asked if this was supported at https://github.com/jenkinsci/configuration-as-code-plugin/pull/183#issuecomment-450362458, and @ndeloof confirmed it and pointed me to the docs. Is this issue about something else? Appreciate any clarification, thanks in advance.

**Comment by casz** [comment]
@antgel we are talking about string interpolation https://en.wikipedia.org/wiki/String_interpolation#Examples

This is supported
```
  mySecret: ${aSecret}
```

This is not supported and this is also string interpolation :smile:
```
  myOtherSecretInAstring: http://mydomain.com/?token=${aSecret}
```

**Comment by antgel** [comment]
@casz Thanks for the clarity. In https://github.com/jenkinsci/configuration-as-code-plugin/pull/183#issuecomment-451398385 I thought I explained that was our use case, but perhaps it wasn't clear. So we are pretty much at the point where we need this, to differentiate between cloud regions and terraform environments. @dgarzon any way we can fund the development to accelerate this?

**Comment by dgarzon** [comment]
@antgel no need. We also need this functionality at my current company, and they are sponsoring this work! I'll do my best to have something working between today and Monday.

**Comment by ndeloof** [comment]
Agree. secret replacement is [limited today](https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/plugin/src/main/java/io/jenkins/plugins/casc/SecretSource.java#L22), we planned to re-implement it a stronger way, with an actual String interpolation library. If you know a nice one, let us know.

**Comment by dgarzon** [comment]
@ndeloof I think this is the one we should use: https://github.com/lantunes/interpolatd/blob/master/README.md

Went over the README and it is very simple, yet powerful. What do you think? If you give thumbs up I can have a working implementation today for sure.

**Comment by casz** [comment]
@dgarzon 
The library seems like a good fit, the enclosure should still be `${}` and the option to escape it would be nice.

I have a use case for having secrets being escaped currently that is done with a space character which feels wrong.

but also the use case for string interpolation :)

**Comment by ndeloof** [comment]
looks like a perfect candidate, especially having no third party dependencies

**Comment by dgarzon** [comment]
Perfect! I'll work on it today :)

**Comment by dgarzon** [comment]
@rdpa @casz @ndeloof @antgel I have implemented this feature here: https://github.com/jenkinsci/configuration-as-code-plugin/pull/699

**Comment by rdpa** [comment]
Awesome, thank you!

**Comment by casz** [comment]
fixed in #699 and #749

**Comment by benfiedler** [comment]
This may be a dumb question but does the interpolation now work for environment variables too? They suffered from the same limitation of having to be the entire value of the  key-value pair as well.

**Comment by casz** [comment]
@benfiedler string interpolation is fixed in `v1.8`

See https://github.com/jenkinsci/configuration-as-code-plugin#handling-secrets for how to use it.
